src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+    data_.Jsum_dx     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jsum_xnext  = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jsum_dxnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Jdiff_xnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jdiff_x     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Jg_dx     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jg_dxnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jg_u      = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Ldx   = Eigen::VectorXd::Zero(ndx_);
+    data_.Ldxdx = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Ldxu  = Eigen::MatrixXd::Zero(ndx_, nu_);
+}
+
+MultipleShootingNlp::~MultipleShootingNlp() {}
+
+bool MultipleShootingNlp::get_nlp_info(Ipopt::Index   &n,
+                                       Ipopt::Index   &m,
+                                       Ipopt::Index   &nnz_jac_g,
+                                       Ipopt::Index   &nnz_h_lag,
+                                       IndexStyleEnum &index_style)
+{
+    n = nvar_;
+
+    m = nconst_;
+
+    // Jacobian nonzeros for dynamic constraints
+    nnz_jac_g = T_ * ndx_ * (2 * ndx_ + nu_);
+
+    // Jacobian nonzeros for initial condition
+    nnz_jac_g += ndx_;
+
+    // Hessian is only affected by costs
+    // Running Costs
+    std::size_t nonzero = 0;
+    for (size_t i = 1; i <= (ndx_ + nu_); i++) {
+        nonzero += i;
+    }
+    nnz_h_lag = T_ * nonzero;
+
+    // Terminal Costs
+    nonzero = 0;
+    for (size_t i = 1; i <= ndx_; i++) {
+        nonzero += i;
+    }
+    nnz_h_lag += nonzero;
+
+    // use the C style indexing (0-based)
+    index_style = Ipopt::TNLP::C_STYLE;
+
+    return true;
+}
+// [TNLP_get_nlp_info]
+
+// [TNLP_get_bounds_info]
+// returns the variable bounds
+bool MultipleShootingNlp::get_bounds_info(Ipopt::Index   n,
+                                          Ipopt::Number *x_l,
+                                          Ipopt::Number *x_u,
+                                          Ipopt::Index   m,
+                                          Ipopt::Number *g_l,
+                                          Ipopt::Number *g_u)
+{
+    // here, the n and m we gave IPOPT in get_nlp_info are passed back to us.
+    // If desired, we could assert to make sure they are what we think they are.
+    assert(n == nvar_);
+    assert(m == nconst_);
+
+    // // the variables have lower bounds of 1
+    // for (Ipopt::Index i = 0; i < n; i++) {
+    //     x_l[i] = -2e19;
+    //     x_u[i] = 2e19;
+    // }
+
+    // Adding bounds
+    for (size_t i = 0; i < T_; i++) {
+        // Running state bounds
+        for (size_t j = 0; j < ndx_; j++) {
+            x_l[i * (ndx_ + nu_) + j] = -2e19;
+            x_u[i * (ndx_ + nu_) + j] = 2e19;
+        }
+        // Control bounds
+        for (size_t j = 0; j < nu_; j++) {
+            x_l[i * (ndx_ + nu_) + ndx_ + j] = problem_->get_runningModels()[i]->get_has_control_limits()
```

Comment:
Simplify the code as
```suggestion
    const boost::shared_ptr<ActionModelAbstract>& model = problem_->get_runningModels()[i];
    const boost::shared_ptr<ActionDataAbstract>& data = problem_->get_runningDatas()[i];
            x_l[i * (ndx_ + nu_) + ndx_ + j] = model->get_has_control_limits()
```

Do this in other case. I won't mark them.